2022/day9/Program.fs

```diff
@@ -0,0 +1,91 @@
+﻿// https://adventofcode.com/2022/day/9
+open System.IO
+
+type Pos = struct
+    val X: int
+    val Y: int
+    new (x: int, y: int) = { X = x; Y = y }
+end
+
+// Create a new position by moving from the previous position.
+let movePos (x: int) (y: int) (p: Pos) = Pos(p.X + x, p.Y + y)
+
+// Parse a move sequence into a sequence of move function calls.
+// This will convert multi-cell moves into a sequence 1-cell of moves.
+let parseMoveSeq (move: string) =
+    match move.ToUpper().Split(' ') with
+    | [| "U"; count |] -> seq { for c in 1 .. (int count) -> movePos 0 -1 }
+    | [| "D"; count |] -> seq { for c in 1 .. (int count) -> movePos 0 +1 }
+    | [| "L"; count |] -> seq { for c in 1 .. (int count) -> movePos -1 0 }
+    | [| "R"; count |] -> seq { for c in 1 .. (int count) -> movePos +1 0 }
+    | _ -> []
+
+// Pull from the previous knot in the rope. If they are within 1 cell in
+// any direction, including diagonally, then no movement is needed.
+// Otherwise, move to follow the head, moving 0 or 1 position on each axis.
+let pullPos (h: Pos) (t: Pos) : Pos =
+    match (h.X - t.X, h.Y - t.Y) with
+    | (x,y) when abs x <= 1 && abs y <= 1 -> t
+    | (x,y) -> movePos (x.CompareTo(0)) (y.CompareTo(0)) t
+
+let processRopeMoves (knotCount: int) (verbosity: int) =
+    // Create the rope with all knot positions initialized to (0, 0).
+    let mutable knots = Array.create knotCount (Pos(0, 0))
+
+    // Track the visits for all knots in the rope.
+    let mutable visits = Array.create knotCount [Pos (0, 0)]
+
+    // For each line in the input file, parse the line into a sequence of moves.
+    // Then execute those moves and track the visits of every knot in the rope.
+    // https://adventofcode.com/2022/day/9/input
+    for line in File.ReadAllLines("input.txt") do
+        let (H, T) = (Array.head knots, Array.last knots)
+
+        if verbosity > 1 then
+            printfn "MOVE \tHEAD         \tVisits\tTAIL         \tVisits"
+            printfn "%-5s\t(%+5d,%+5d)\t%-6d\t(%+5d,%+5d)\t%-6d" line H.X H.Y (List.distinct (Array.head visits)).Length T.X T.Y (List.distinct (Array.last visits)).Length
```

Comment:
This line feels a bit long and cryptic.
But what I really want to point out is how a piece of code like this
`(List.distinct (Array.head visits))`
Is hard to process in the head, because it reads backwards. You first pick the head of visits and then distinct on it. In F# we have the pipe operator `|` I'm sure you've seen it being used in multiline parsing, but I would recommend it sometimes in just inline situation like this:
`visits |> Array.head |> List.distinct`
I think that reads more easily (at least IMO), take visits, get the head, then List.distinct. From left to right.
There is unfortunately not a support for .Length at the end (right now) so you can't do this:
`visits |> Array.head |> List.distinct |> .Length`
but it should come in the future, for now you need to do this, thankfully this is a List that we are dealing with,
so instead of writing:
(visits |> Array.head |> List.distinct).Length
or this:
`visits |> Array.head |> List.distinct |> fun x -> x.Length`
We have already named function for getting Length:
`visits |> Array.head |> List.distinct |> List.length`